tables/audisp/src/fileeventstableplugin.cpp

```diff
@@ -0,0 +1,241 @@
+    obj.reset(ptr);
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
+
+  } catch (const Status &status) {
+    return status;
+  }
+}
+
+FileEventsTablePlugin::~FileEventsTablePlugin() {}
+
+const std::string &FileEventsTablePlugin::name() const {
+  static const std::string kTableName{"file_events"};
+
+  return kTableName;
+}
+
+const FileEventsTablePlugin::Schema &FileEventsTablePlugin::schema() const {
+  static const Schema kTableSchema = {
+      {"syscall", IVirtualTable::ColumnType::String},
+      {"pid", IVirtualTable::ColumnType::Integer},
+      {"ppid", IVirtualTable::ColumnType::Integer},
+      {"uid", IVirtualTable::ColumnType::Integer},
+      {"gid", IVirtualTable::ColumnType::Integer},
+      {"auid", IVirtualTable::ColumnType::Integer},
+      {"euid", IVirtualTable::ColumnType::Integer},
+      {"egid", IVirtualTable::ColumnType::Integer},
+      {"exe", IVirtualTable::ColumnType::String},
+      {"path", IVirtualTable::ColumnType::String},
+      {"inode", IVirtualTable::ColumnType::Integer},
+      {"time", IVirtualTable::ColumnType::Integer}};
+
+  return kTableSchema;
+}
+
+Status FileEventsTablePlugin::generateRowList(RowList &row_list) {
+  std::lock_guard<std::mutex> lock(d->row_list_mutex);
+
+  row_list = std::move(d->row_list);
+  d->row_list = {};
+
+  return Status::success();
+}
+
+Status FileEventsTablePlugin::processEvents(
+    const IAudispConsumer::AuditEventList &event_list) {
+  RowList generated_row_list;
+
+  for (const auto &audit_event : event_list) {
+    Row row;
+
+    auto status = generateRow(row, audit_event);
+    if (!status.succeeded()) {
+      return status;
+    }
+
+    if (!row.empty()) {
+      {
+        std::lock_guard<std::mutex> lock(d->row_list_mutex);
+        d->row_list.push_back(row);
+      }
+    }
+  }
+
+  if (d->row_list.size() > d->max_queued_row_count) {
+
+    std::lock_guard<std::mutex> lock(d->row_list_mutex);
+
+    if (d->row_list.size() > d->max_queued_row_count) {
+      auto rows_to_remove = d->row_list.size() - d->max_queued_row_count;
+
+      d->logger.logMessage(IZeekLogger::Severity::Warning,
+                           "file_events: Dropping " +
+                               std::to_string(rows_to_remove) +
+                               " rows (max row count is set to " +
+                               std::to_string(d->max_queued_row_count) + ")");
+
+      d->row_list.erase(d->row_list.begin(),
+                        std::next(d->row_list.begin(), rows_to_remove));
+    }
+  }
+
+  return Status::success();
+}
+
+FileEventsTablePlugin::FileEventsTablePlugin(IZeekConfiguration &configuration,
+                                             IZeekLogger &logger)
+    : d(new PrivateData(configuration, logger)) {
+
+  d->max_queued_row_count = d->configuration.maxQueuedRowCount();
+}
+
+std::string FileEventsTablePlugin::CombinePaths(const std::string &cwd,
```

Comment:
I am confused why we do all this extra manipulation involving quotes here. Is that an implementation detail or an aspect users would care about.

Using `std::filesystem::path`, is this function equivalent to the following?
```.cpp
std::filesystem::path CombinePaths(const std::filesystem::path &cwd,
                                   const std::filesystem::path &path) {
  if (path.is_absolute())
    return path;

  return cwd / path;
}
```

Reply:
This is how AUDIT_PATH events are usually delivered by Audit. Quote decoding could be moved inside audit_utils.cpp and its unit tests implemented here: https://github.com/zeek/zeek-agent/blob/master/components/zeekaudisp/tests/audit_utils.cpp

Reply:
This function `CombinePaths` is already exercised by tests in `tables/audisp/tests/fileeventstableplugin.cpp`. However, I will add one more test in the next commit to cover one additional basic block in that function

Reply:
Added a new test case so marking this resolve.